### ISSUE TITLE
IPC::Open3 open3() failure problem

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#4056]
Migrated from [rt.perl.org#7072](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=7072) (status was 'resolved')

Searchable as RT7072$

**Comment by p5pRT** [comment]
### From james@snowy.squish.net



IPC&#8203;::Open3\, after forking\, attempts an exec in the child\.  If the
exec fails\, IPC&#8203;::Open3 calls croak\.  This has the unfortunate
side affect of running all the END blocks\.  What makes this
more annoying is that due to being after a fork\, this could happen
several lines after the open3 call\, undeterministically at a
random place\, and since stdin/stdout/stderr have been redirected\,
potentially hard to notice that this is what is happening\.

The typical scenario of opening tmp files with END blocks to
clean up unexpected situations falls foul of open3 when for no
apparent reason your temporary files vanish for apparently no
reason\.

I would like to know how you are supposed to use open3 in such a
way as to support END blocks and being able to cope with exec
failing\.

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=low

Site configuration information for perl v5.6.0:

Configured by root at Mon Jun 19 19:59:32 BST 2000.

Summary of my perl5 (revision 5.0 version 6 subversion 0) configuration:
  Platform:
    osname=openbsd, osvers=2.7, archname=i386-openbsd
    uname='openbsd'
    config_args='-Dopenbsd_distribution=defined -dsE'
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=undef use5005threads=undef useithreads=undef usemultiplicity=undef
    useperlio=undef d_sfio=undef uselargefiles=define 
    use64bitint=undef use64bitall=undef uselongdouble=undef usesocks=undef
  Compiler:
    cc='cc', optimize='-O2', gccversion=2.95.2 19991024 (release)
    cppflags='-fno-strict-aliasing -I/usr/local/include'
    ccflags ='-fno-strict-aliasing -I/usr/local/include'
    stdchar='char', d_stdstdio=undef, usevfork=true
    intsize=4, longsize=4, ptrsize=4, doublesize=8
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=4, usemymalloc=n, prototype=define
  Linker and Libraries:
    ld='ld', ldflags =''
    libpth=/usr/lib
    libs=-lm -lc
    libc=/usr/lib/libc.so.25.0, so=so, useshrplib=true, libperl=libperl.so.6.0
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=define, ccdlflags=' '
    cccdlflags='-DPIC -fPIC ', lddlflags='-Bshareable '

Locally applied patches:
    


@INC for perl v5.6.0:
    /usr/libdata/perl5/i386-openbsd/5.6.0
    /usr/local/libdata/perl5/i386-openbsd/5.6.0
    /usr/libdata/perl5
    /usr/local/libdata/perl5
    /usr/local/libdata/perl5/site_perl/i386-openbsd
    /usr/libdata/perl5/site_perl/i386-openbsd
    /usr/local/libdata/perl5/site_perl
    /usr/libdata/perl5/site_perl
    /usr/lib/perl5/site_perl
    .


Environment for perl v5.6.0:
    HOME=/home/james
    LANG (unset)
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/home/james/bin:/home/james/bin:/home/james/bin:/sbin:/usr/sbin:/bin:/usr/bin:/usr/local/sbin:/usr/local/bin
    PERL_BADLANG (unset)
    SHELL=/usr/local/bin/bash

```
</details>

**Comment by p5pRT** [comment]
### From p5p@spam.wizbit.be

This was resolved \(a long time ago\) by using POSIX&#8203;::\_exit\(255\) in 
IPC&#8203;::Open3 which does not execute the END block\.

**Comment by p5pRT** [comment]
p5p@spam.wizbit.be - Status changed from 'open' to 'resolved'
